qmi/core/transport.py

```diff
@@ -858,70 +853,105 @@ def read(self, nbytes: int, timeout: Optional[float]) -> bytes:
             ~qmi.core.exceptions.QMI_EndOfInputException: If an end-of-message indicator
                 is received before the requested number of bytes is reached.
         """
-
         self._check_is_open()
-
         # USB requires a timeout
         if timeout is None:
             timeout = self.DEFAULT_READ_TIMEOUT
 
-        if not self._read_buffer:
-            # Read buffer is empty - read a new message from the instrument.
-            self._read_buffer = self._read_message(timeout)
-
-        # Check that the buffer contains enough data.
-        if len(self._read_buffer) < nbytes:
-            raise QMI_EndOfInputException("Expecting {} bytes but only {} bytes remaining in USBTMC message"
-                                          .format(nbytes, len(self._read_buffer)))
+        nbuf = len(self._read_buffer)
```

Comment:
I don't understand what the difference is between the removed and added code? Both seem to read message until timeout, and throw error if length is not what is expected?

Reply:
but this part of the code will throw an exception if the first read doesn't return the correct amount of bytes:
```python
    if nbuf != nbytes:
        _logger.debug("USBTMC read buffer contained data %s" % self._read_buffer.decode())
        self._read_buffer = bytes()
        raise QMI_EndOfInputException(
            f"The read buffer did not contain expected bytes of data ({nbuf} != {nbytes}."
        )
```

Reply:
That is correct. The `while True:` indeed is quite useless. I will remove it.

What comes to the other changes is that I changed the order to return the data from the buffer directly if there is exactly the correct amount of bytes in. Any wrong number of data, or no data, should result in reading a new message. What I realize here now is that 
a) The read buffer should be at that point cleared as the USB should only return whole messages and it should not be possible(?) that only partial messages are returned unless TimeoutError is raised. 
b) the new message read in should not be added to existing read buffer, but checked directly if it has the expected amount of  bytes.

I think I had the original `while` loop idea as being for cases where less than expected `nbytes` data has been delivered by (multiple) read(s) and it would loop and add the bytes into the read buffer until at least `nbytes` are available, unless it gets some exception. But the function can only do a partial readout only in case of a timeout (as USB should always return one whole message), which means that an exception is raised and the `while` loop will be broken - so the re-read and adding to the `_read_buffer` cannot work. But I then somehow forgot the whole thing and didn't remove it.

Reply:
Perhaps saving the read into the `_read_buffer`, and then calling to read again could be a solution, but the correct way would rather be to increase the timeout such that the whole message will be read in and not only partially.

Reply:
If going for the solution of increasing timeout, we can actually remove the whole `self._read_buffer` as this function was the only one actually adding data to it. If require rather that the user uses longer timeout next time instead of allowing it to place several read calls in sequence, we can remove it. Do you agree?

Reply:
Yep. let's go with removing the `_read_buffer`.  That simplifies things a bit and keeps it closer to how usbtmc works in reality. Then the user has multiple options. Either read a message it knows how long it will be, or just read a random message with an appropriate timeout.

This does make `discard_read` pointless again? So we can either have that just be pass, or raise not implemented?

Reply:
`discard_read` can still try to read from the instrument and discard if it got something as a reply.